providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/FlagdProvider.java

```diff
@@ -76,18 +139,91 @@ public FlagdProvider(String socketPath) {
      *
      */
     public FlagdProvider(String host, int port, boolean tls, String certPath) {
-        this(buildServiceStub(host, port, tls, certPath, null));
+        this(
+            buildServiceBlockingStub(host, port, tls, certPath, null),
+            buildServiceStub(host, port, tls, certPath, null),
+            fallBackToEnvOrDefault(CACHE_ENV_VAR_NAME, DEFAULT_CACHE),
+            fallBackToEnvOrDefault(MAX_CACHE_SIZE_ENV_VAR_NAME, DEFAULT_MAX_CACHE_SIZE),
+            fallBackToEnvOrDefault(MAX_EVENT_STREAM_RETRIES_ENV_VAR_NAME, DEFAULT_MAX_EVENT_STREAM_RETRIES)
+        );
+    }
+
+    /**
+     * Create a new FlagdProvider instance.
+     *
+     * @param host                  flagd server host, defaults to "localhost"
+     * @param port                  flagd server port, defaults to 8013
+     * @param tls                   use TLS, defaults to false
+     * @param certPath              path for server certificate, defaults to null to, using
+     *                              system certs
+     * @param cache                 caching implementation to use (lru)
+     * @param maxCacheSize          limit of the number of cached values for each type of flag
+     * @param maxEventStreamRetries limit of the number of attempts to connect to flagd's event stream,
+     *                              on successful connection the attempts are reset
+     *
+     */
+    public FlagdProvider(String host, int port, boolean tls, String certPath, String cache,
+        int maxCacheSize, int maxEventStreamRetries) {
+        this(
+            buildServiceBlockingStub(host, port, tls, certPath, null),
+            buildServiceStub(host, port, tls, certPath, null),
+            cache, maxCacheSize, maxEventStreamRetries
+        );
     }
 
     /**
      * Create a new FlagdProvider instance.
      */
     public FlagdProvider() {
-        this(buildServiceStub(null, null, null, null, null));
+        this(
+            buildServiceBlockingStub(null, null, null, null, null),
+            buildServiceStub(null, null, null, null, null),
+            fallBackToEnvOrDefault(CACHE_ENV_VAR_NAME, DEFAULT_CACHE),
+            fallBackToEnvOrDefault(MAX_CACHE_SIZE_ENV_VAR_NAME, DEFAULT_MAX_CACHE_SIZE),
+            fallBackToEnvOrDefault(MAX_EVENT_STREAM_RETRIES_ENV_VAR_NAME, DEFAULT_MAX_EVENT_STREAM_RETRIES)
+        );
     }
 
-    FlagdProvider(ServiceBlockingStub serviceStub) {
+    FlagdProvider(ServiceBlockingStub serviceBlockingStub, ServiceStub serviceStub, String cache,
+        int maxCacheSize, int maxEventStreamRetries) {
+        this.serviceBlockingStub = serviceBlockingStub;
         this.serviceStub = serviceStub;
+        this.eventStreamAlive = false;
+        if (cache != null) {
+            initCache(cache, maxCacheSize);
+        }
+        this.maxEventStreamRetries = maxEventStreamRetries;
+        this.handleEvents();
+    }
+
+    private void initCache(String cache, int maxSize) {
+        switch (cache) {
```

Comment:
You can use fallthrough behavior and avoid the recursive call :) 

Something like below should do, 
```
    switch (cache) {
           case DISABLED:
                return;
            case LRU_CACHE:
            default:
                 this.booleanCache = Collections.synchronizedMap(
                    new LRUMap<String, ProviderEvaluation<Boolean>>(maxSize));
                this.stringCache = Collections.synchronizedMap(new LRUMap<String, ProviderEvaluation<String>>(maxSize));
                this.doubleCache = Collections.synchronizedMap(new LRUMap<String, ProviderEvaluation<Double>>(maxSize));
                this.integerCache = Collections.synchronizedMap(
                    new LRUMap<String, ProviderEvaluation<Integer>>(maxSize));
                this.objectCache = Collections.synchronizedMap(new LRUMap<String, ProviderEvaluation<Value>>(maxSize));
                return;
    }
  
  ```

Reply:
Nice suggestion, done